pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -1445,7 +1445,88 @@ func (p *Provider) Create(
 			return err
 		}
 	}
-	return propagateDiskLabels(l, project, labels, zoneToHostNames, &opts)
+	return propagateDiskLabels(l, project, labels, zoneToHostNames, opts.SSDOpts.UseLocalSSD)
+}
+
+// computeGrowDistribution computes the distribution of new nodes across the
+// existing instance groups. The distribution is computed naively, for
+// simplicity, by growing the instance group with the smallest size, and then
+// the next smallest, and so on.
+func computeGrowDistribution(groups []jsonManagedInstanceGroup, newNodeCount int) []int {
+	addCount := make([]int, len(groups))
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].Size < groups[j].Size
+	})
+	curIndex := 0
+	for i := 0; i < newNodeCount; i++ {
+		nextIndex := (curIndex + 1) % len(groups)
+		if groups[curIndex].Size+addCount[curIndex] >
+			groups[nextIndex].Size+addCount[nextIndex] {
+			curIndex++
+			if curIndex >= len(groups) {
+				curIndex = 0
+			}
+		} else {
+			curIndex = 0
+		}
+		addCount[curIndex]++
+	}
+	return addCount
+}
+
+func (p *Provider) Grow(l *logger.Logger, vms vm.List, clusterName string, names []string) error {
+	project := vms[0].Project
+	groupName := instanceGroupName(clusterName)
+	groups, err := listManagedInstanceGroups(project, groupName)
+	if err != nil {
+		return err
+	}
+
+	newNodeCount := len(names)
+	addCounts := computeGrowDistribution(groups, newNodeCount)
+
+	zoneToHostNames := make(map[string][]string)
+	var g errgroup.Group
+	for idx, group := range groups {
+		addCount := addCounts[idx]
+		if addCount == 0 {
+			continue
+		}
+		createArgs := []string{"compute", "instance-groups", "managed", "create-instance", "--zone", group.Zone, groupName}
+		for i := 0; i < addCount; i++ {
+			name := names[i]
+			names = names[1:]
```

Comment:
It makes sense that we remove the first element of `names`, that vm has been added. But in that case, shouldn't the name we are adding always be `name[0]`?

i.e. we have the instance groups with size [1, 1, 1]

We call Grow(4).

`names = [vm-1, vm-2, vm-3, vm-4]`

Iteration 1: `name=vm-1`, `names=[vm-2, vm-3, vm-4]`

Iteration 2:  `name=vm-3`, `names=[vm-3, vm-4]`, which is wrong since we removed `vm-2` but didn't actually add it yet.